### ISSUE TITLE
[Questions] blitImage performance and BLImage color modulation

### ISSUE THREAD

**Issue by iryont** [blend2d/blend2d#62]
I wrote a simplistic backend for my game using Blend2D. However, its performance with pretty much only blitImage being called thousands of times with images of varying width from 16x16 to 64x64 (and sometimes bigger for UI elements, but only a few) is rather bad compared to simplistic software blitter on SDL2 (about 38 fps on Blend2D vs 76 fps on SDL2 software renderer). I kind of expected Blend2D to a bit faster due to all the SIMD going on unlike within SDL2 software blitter. However, it does not appear to be the case.

The question would be, is this rather expected because Blend2D seems to be vector graphics engine and I shouldn't expect much performance from simplistic image blitting or is there something I can do to make it faster?

Another question - how can I modulate the color of BLImage with blitImage function? Is this even possible? For example, I have a white mask which needs to be first modulated by a color value and then blended into the target with specific composition mode (setCompOp). I did try doing it with **setStyle(BL_COMP_OP_MODULATE, BLRgba32(255, 0, 0, 255))**, but it does not appear to be working.

**Comment by kobalicek** [comment]
1.

If you do pixel aligned blits Blend2D is not expected to be slower than other library. However, a library that utilizes AVX2 may actually be faster as Blend2D is only using 128-bit SSE/AVX at the moment and it would basically be matching your benchmark.

I would just note that image blitting is usually a fast-path operation. There are specialized blitters in all 2D libraries that do aligned blits and it's usually the most optimized operation in such libraries together with solid fills. With Blend2D we focused primarily on path rasterization and rendering, so there is definitely a room for improvement in blitting scenarios - the first thing would be to utilize AVX2 when available, and the second optimization would be multithreading, which I'm working on at the moment.

2.

You mean masked blitting? Using two images, one to blit and another to be used as mask? This is currently not supported although planned feature.

**Comment by iryont** [comment]
1.

Actually I have a menu in which I do some scaling of background image (2048 x 1152) which I do scale to fit the application. In my case it's full desktop mode in 3840 x 1600. The rest images are pretty much blitted without any scaling whatsoever. In this case I do get 28 fps which is really low. Removing the background which is only one image being scaled gets me up to 180-190 fps. The same goes to game which has a framebuffer on its own BLContext which is later on scaled. It seems like linear / bilinear filtering does take a lot of power, so there is any possible for BLImage to use nearest pixel scaling with blitImage function? I would love to turn off scaling for certain BLImage which isn't a BLPattern for that matter. SDL2 uses nearest neighbor interpolation which is fine for a game like that and is pretty fast even despite the fact that blitting functions are not vectorized (even by the compiler).

2.

No, I mean the color (ARGB) modulation of the BLImage. Let me give an example, below are three images with different composition modes with the background:

![b1](https://i.imgur.com/pN9RQ4o.png)

And here is with red (0xFF00000000) modulation:

![b2](https://i.imgur.com/IzYuNRa.png)

Basically just applying a color to BLImage before blending it in the context with selected composition mode.

**Comment by iryont** [comment]
@kobalicek 

I do apologize for mentioning you here, but the color modulation (with alpha modulation) of BLImage is an important feature for me. That being the case, is there any possibility for such thing to be actually implemented? That is an important aspect of my rendering engine for a 2D game and I honestly wanted to have Blend2D as a good software fallback for certain users. I did actually write the Blend2D backend already and it would be a shame for it to go to waste after all.

**Comment by kobalicek** [comment]
I will definitely look into this, but I think this feature is non-standard in general and I would like to have this implemented through pixel-effect pipeline, which is not high priority at the moment. I also though about parametrized compositors, which would also allow something like this, most likely, but that's also not worked on at the moment.

I think additionally this could be implemented with layers as well, which is something that is definitely planned - in that case you would create a temporary layer, blit image, and fill color with `BL_COMP_OP_MODULATE`, and then end the layer and blit it on the topmost one.

Please be patient, we definitely want more features, but our time is also limited and we have to prioritize.

**Comment by iryont** [comment]
I actually thought about the idea with framebuffer (another BLContext), but it immediately looked like a perfect performance killer for that matter. Especially since I render glyphs from BLImage and there are a lot of those.

I honestly thought this feature is rather standard in any rendering pipeline for that matter. I mean, e.g. glColor in OpenGL does modulate textures, same goes to SetRenderState(D3DRS_TEXTUREFACTOR, D3DCOLOR_ARGB(X,Y,Z,W)) in DirectX 9. The same goes to SDL2 which has a software renderer on its own:

https://wiki.libsdl.org/SDL_SetSurfaceColorMod

It would also have to be efficient enough so it doesn't waste much if any computing power for that matter.

**Comment by iryont** [comment]
Interestingly enough I decided to try the approach with different BLContext to see what kind of penalty it will bring. I did render all the glyphs into sub-context with attached temporary usable BLImage and then modulated them with BL_COMP_OP_MODULATE and fillRect with specified color inside that context. Then I did blit the image into the main context. As expected this approach works and I did not lose many fps as I did assume before. In fact, being able to hit about 61 fps without any modulation I do get about 57 fps with this approach and fully modulated textures. That sounds more than acceptable, especially since without bilinear filtering I am still able to get about 100 fps.

I suppose I will be using Blend2D after all :)
